src/hotspot/share/runtime/vmThread.hpp

```diff
@@ -143,16 +99,24 @@ class VMThread: public NamedThread {
   static void execute(VM_Operation* op);
 
   // Returns the current vm operation if any.
-  static VM_Operation* vm_operation()             { return _cur_vm_operation; }
-  static VM_Operation::VMOp_Type vm_op_type()     { return _cur_vm_operation->type(); }
+  static VM_Operation* vm_operation()             {
+    assert(Thread::current()->is_VM_thread(), "Must be");
+    return _cur_vm_operation;
+  }
+  static VM_Operation::VMOp_Type vm_op_type()     {
+    assert(Thread::current()->is_VM_thread(), "Must be");
+    return _cur_vm_operation->type();
```

Comment:
These two might just delegate and check a bit more thoroughly.

```
  static VM_Operation* vm_operation()             {
  static VM_Operation::VMOp_Type vm_op_type()     { return _cur_vm_operation->type(); }	    assert(Thread::current()->is_VM_thread(), "Must be");
    return _cur_vm_operation;
  }

  static VM_Operation::VMOp_Type vm_op_type()     {
    VM_Operation* op = vm_operation();
    assert (op != NULL, "sanity");
    return op->type();
  }
```

Haven't checked if `vm_op_type` helps readability at actual usages, maybe it is unnecessary to begin with.

Reply:
vm_operation() may return NULL as callers handles this, as long it's the VM Thread calling a non NULL is also fine.
But we should remove this method and do something else, but for another PR :)

vm_op_type() is only used at one place, but I think your version is much better so I took that!